Improve the RPS and add test logs to the script for Merino Load testing (#283)
mozilla-services/merino: test-engineering/load-tests/README.md

### For locust master and worker nodes

- `LOAD_TESTS__LOGGING_LEVEL`: Level for the logger in the load tests as an int (`10` for `DEBUG`, `20` for `INFO` etc.)
- `KINTO__SERVER_URL`: Server URL of the Kinto instance to download suggestions from
- `KINTO__BUCKET`: Kinto bucket with the suggestions

> - `KINTO__COLLECTION`: Kinto collection with the suggestions

Author: It doesn't look like my changes. I think something got mixed up when I pulled the latest changes on my branch as well as main merged. I will revert it.

Make the file executable by:

```text

> chmod +x setup_k8.sh

Reviewer: I assume `k8` is supposed to stand for Kubernetes here? If so, this should be `k8s`.

See https://github.com/kubernetes/kubernetes#kubernetes-k8s

```

Then, run the file:

```text
./setup_k8.sh
```
Select the option `create`, it goes through the process of creating a cluster, setting up the env variables
and building the docker image.

> *Note*: Be sure to delete the cluster(steps mentioned at the end of this doc) after you are done running load tests and copying the logs.

Reviewer: ```suggestion
**Note:** Be sure to delete the cluster (steps mentioned at the end of this doc) after you are done running load tests and copying the logs.
```

Currently, it's set to run `10` workers but it can be changed by running above command.

# How to grab the logs

Get the master pod name by running:

```text
kubectl get pods -o wide
```

then run:

```text
kubectl cp <master-pod-name>:/home/locust/merino_stats.csv merino_stats.csv
kubectl cp <master-pod-name>:/home/locust/merino_exceptions.csv merino_exceptions.csv
kubectl cp <master-pod-name>:/home/locust/merino_failures.csv merino_failures.csv
```

> To remove all the GET requests and retain only the aggregate summary use:

Reviewer: ```suggestion
To remove all the `GET` requests and retain only the aggregate summary use:
```

```text
cat merino_stats.csv | grep -Ev "^GET," > merino_stats.csv.tmp
mv merino_stats.csv.tmp merino_stats.csv
```

> Thereafter, copy these 3 files(merino_exceptions.csv, merino_failures.csv, merino_summary.csv) to [gist](https://gist.github.com/) and provide a link to it in the [Merino load test history doc](https://docs.google.com/document/d/1BGNhKuclUH40Bit9KxYWLiv_N_VnE66uxi9pBFbRWbg/edit)

Reviewer 2: ```suggestion
Thereafter, copy these 3 files (merino_stats.csv, merino_exceptions.csv, merino_failures.csv) to [gist](https://gist.github.com/new) and provide a link to it in the [Merino load test history doc](https://docs.google.com/document/d/1BGNhKuclUH40Bit9KxYWLiv_N_VnE66uxi9pBFbRWbg/edit)
```

> Thereafter, copy these 3 files (merino_stats.csv, merino_exceptions.csv, merino_failures.csv) to [gist](https://gist.github.com/new) and provide a link to it in the [Merino load test history doc](https://docs.google.com/document/d/1BGNhKuclUH40Bit9KxYWLiv_N_VnE66uxi9pBFbRWbg/edit)

Reviewer: ```suggestion
Thereafter, copy these 3 files (`merino_stats.csv`, `merino_exceptions.csv`, `merino_failures.csv`) to [gist](https://gist.github.com/new) and provide a link to it in the [Merino load test history doc](https://docs.google.com/document/d/1BGNhKuclUH40Bit9KxYWLiv_N_VnE66uxi9pBFbRWbg/edit)
```

Currently, it's set to run `10` workers but it can be changed by running above command.

# How to grab the logs

Get the master pod name by running:

```text
kubectl get pods -o wide
```

then run:

```text
kubectl cp <master-pod-name>:/home/locust/merino_stats.csv merino_stats.csv
```

```text
kubectl cp <master-pod-name>:/home/locust/merino_exceptions.csv merino_exceptions.csv
```

```text
kubectl cp <master-pod-name>:/home/locust/merino_failures.csv merino_failures.csv
```

To remove all the GET requests and retain only the aggregate summary use:

```text

> cat merino_stats.csv | grep -Ev "^GET," > merino_summary.csv

Reviewer 2: We'd need to test this, but another option is seeing if we just overwrite the merino_stats.csv file with the edited version, instead of adding a 4th or 5th file (well, I guess fifth on the **master** node, fourth on the local machine since we don't copy all .csv files from the pod).
But I'm not 100% certain we can <kbd>cat</kbd> a file and pipe it and then redirect it back to itself. It doesn't always work for me and sometimes you're just left with an empty file.

Reviewer 3: I'm sure that we **cannot** cat a file and the redirect it back to itself. The shell truncates the file before executing the command. You could use a temporary file for this

```suggestion
cat merino_stats.csv | grep -Ev "^GET," > merino_stats.csv.tmp
mv merino_stats.csv.tmp merino_stats.csv
```
